staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go

```diff
@@ -116,3 +120,116 @@ func CRDRemoveFinalizer(crd *CustomResourceDefinition, needle string) {
 	}
 	crd.Finalizers = newFinalizers
 }
+
+// HasCRDVersion returns true if `version` is in the list of CRD's versions.
+func HasCRDVersion(crd *CustomResourceDefinition, version string) bool {
+	for _, v := range crd.Spec.Versions {
+		if v.Name == version {
+			return true
+		}
+	}
+	return false
+}
+
+// GetCRDStorageVersion returns the storage version for given CRD.
+func GetCRDStorageVersion(crd *CustomResourceDefinition) (string, error) {
+	for _, v := range crd.Spec.Versions {
+		if v.Storage {
+			return v.Name, nil
+		}
+	}
+	// This should not happened if crd is valid
+	return "", fmt.Errorf("invalid CustomResourceDefinition, no storage version")
+}
+
+func IsStoredVersion(crd *CustomResourceDefinition, version string) bool {
+	for _, v := range crd.Status.StoredVersions {
+		if version == v {
+			return true
+		}
+	}
+	return false
+}
+
+func OrderedVersionListFromCRDs(crds []*CustomResourceDefinition, group string) (ret []string) {
+	type crdVersionInfo struct {
+		Name     string
+		versions []CustomResourceDefinitionVersion
+	}
+	var crdInfos []*crdVersionInfo
+	for _, crd := range crds {
+		if crd.Spec.Group != group {
+			continue
+		}
+		crdInfos = append(crdInfos, &crdVersionInfo{crd.Name, crd.Spec.Versions})
+	}
+	sort.SliceStable(crdInfos, func(i, j int) bool {
+		return strings.Compare(crdInfos[i].Name, crdInfos[j].Name) < 0
+	})
+	orderedVersions := list.New()
+	versionsMap := map[string]bool{}
+	for _, info := range crdInfos {
+		for i, v := range info.versions {
+			if !versionsMap[v.Name] {
+				versionsMap[v.Name] = true
+				if orderedVersions.Len() == 0 {
+					orderedVersions.PushBack(v.Name)
+					continue
+				}
+
+				// Try to insert version in the final list where with best effort these two constrains met:
+				// (1) All versions come before this version in this CRD also come before it in the final list
+				// (2) All versions come after this version in this CRD also come after it in the final list
+
+				beforePointer := orderedVersions.Front()
+				count := 0
+				for beforePointer != nil && count < i {
+					for j := 0; j < i; j++ {
+						if beforePointer.Value == info.versions[j].Name {
+							count++
+							break
+						}
+					}
+					beforePointer = beforePointer.Next()
+				}
+
+				if count < i {
+					// We could not satisfy (1)
```

Comment:
// We could not satisfy (1). Do the best to satisfy (2) by pushing it to the front